Add more impls of PartialEq and PartialOrd for strings

joshtriplett opened the pull request:
Currently, some combinations of `&String` and `&str` don't support
comparison operators. For instance, this:

```rust
fn main() {
    let s1 = String::from("hello");
    let s2 = "world";
    _ = s1 < s2;
}
```

will fail with:
```
error[E0308]: mismatched types
 --> src/main.rs:4:14
  |
4 |     _ = s1 < s2;
  |         --   ^^- help: try using a conversion method: `.to_string()`
  |         |    |
  |         |    expected `String`, found `&str`
  |         expected because this is `String`
```

Other combinations only work because the compiler implicitly relies on
impls on different reference types, and that makes such combinations
fragile, breaking if any other impls of `PartialOrd` show up.

Add some additional impls to make such cases work, and to improve
robustness when adding other impls in the future. In particular, I'm hoping that adding these makes it possible to add comparisons with other stringy types without creating as many inference issues.

joshtriplett committed:
Add more impls of PartialEq and PartialOrd for strings

Currently, some combinations of `&String` and `&str` don't support
comparison operators. For instance, this:

```rust
fn main() {
    let s1 = String::from("hello");
    let s2 = "world";
    _ = s1 < s2;
}
```

will fail with:
```
error[E0308]: mismatched types
 --> src/main.rs:4:14
  |
4 |     _ = s1 < s2;
  |         --   ^^- help: try using a conversion method: `.to_string()`
  |         |    |
  |         |    expected `String`, found `&str`
  |         expected because this is `String`
```

Other combinations only work because the compiler implicitly relies on
impls on different reference types, and that makes such combinations
fragile, breaking if any other impls of `PartialOrd` show up.

Add some additional impls to make such cases work, and to improve
robustness when adding other impls in the future.

joshtriplett commented:
@bors try

bors commented:
:hourglass: Trying commit 8b933ddd520a16eec82a67fba7369fbc9da2d4ba with merge 7abdc84c08b19feb1594926bef5a43154769a110...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8b933ddd520a16eec82a67fba7369fbc9da2d4ba","merge_sha":"7abdc84c08b19feb1594926bef5a43154769a110"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12790856050/job/35657602995) [(plain)](https://github.com/rust-lang/rust/commit/8b933ddd520a16eec82a67fba7369fbc9da2d4ba/checks/35657602995/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.rs (revision `all`) ... ok
tests/ui/ref_option/ref_option.all.fixed ... ok

FAILED TEST: tests/ui/iter_overeager_cloned.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e60bfdd447c2b980.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-5813416a2294dc8d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-92ebd8e98d13af54.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-ddb101719ab70c6c.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-e573b0a7912401e3.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-6d39136264af6513.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-2494d295a9e49f01.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-1bd94bc0a5931f31.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-af24b1ebe806a3c8.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-32bdb1831662c978.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-dc8e3506e69419c8.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-012345f38d5585e7.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/iter_overeager_cloned.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/iter_overeager_cloned.stderr` to the actual output
--- tests/ui/iter_overeager_cloned.stderr
+++ <stderr output>
+++ <stderr output>
 error: unnecessarily eager cloning of iterator items
   --> tests/ui/iter_overeager_cloned.rs:12:29
... 26 lines skipped ...
    = help: to override `-D warnings` add `#[allow(clippy::redundant_clone)]`
 
+error: taken reference of right operand
+   |
+   |
+LL |     let _: usize = vec.iter().filter(|x| x == &"2").cloned().count();
+   |                                               |
+   |                                               help: use the right value directly: `"2"`
+   |
+   = note: `-D clippy::op-ref` implied by `-D warnings`
+   = note: `-D clippy::op-ref` implied by `-D warnings`
+   = help: to override `-D warnings` add `#[allow(clippy::op_ref)]`
+
 error: unnecessarily eager cloning of iterator items
   --> tests/ui/iter_overeager_cloned.rs:18:21
... 20 lines skipped ...
    |                                             help: try: `.nth(2).cloned()`
 
+error: taken reference of right operand
+   |
+   |
+LL |     let _ = vec.iter().filter(|x| x == &"2").cloned().nth(2);
+   |                                        |
+   |                                        help: use the right value directly: `"2"`
+
 error: unnecessarily eager cloning of iterator items
 error: unnecessarily eager cloning of iterator items
   --> tests/ui/iter_overeager_cloned.rs:24:13
... 108 lines skipped ...
    |                       help: try: `.any(|x| x.len() == 1)`
-error: aborting due to 19 previous errors
+error: aborting due to 21 previous errors
 




FAILED TEST: tests/ui/op_ref.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e60bfdd447c2b980.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-5813416a2294dc8d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-92ebd8e98d13af54.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-ddb101719ab70c6c.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-e573b0a7912401e3.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-6d39136264af6513.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-2494d295a9e49f01.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-1bd94bc0a5931f31.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-af24b1ebe806a3c8.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-32bdb1831662c978.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-dc8e3506e69419c8.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-012345f38d5585e7.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/op_ref.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/op_ref.stderr` to the actual output
--- tests/ui/op_ref.stderr
+++ <stderr output>
+++ <stderr output>
 error: needlessly taken reference of both operands
   --> tests/ui/op_ref.rs:11:15
... 10 lines skipped ...
 
 error: taken reference of right operand
+   |
+   |
+LL |     if b < &a {
+   |            |
+   |            help: use the right value directly: `a`
+
+
+error: taken reference of right operand
    |
... 19 lines skipped ...
    |                       help: use the right value directly: `three`
 
 
-error: aborting due to 4 previous errors
+error: aborting due to 5 previous errors
 



FAILED TEST: tests/ui/cmp_owned/with_suggestion.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e60bfdd447c2b980.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-5813416a2294dc8d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-92ebd8e98d13af54.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-ddb101719ab70c6c.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-e573b0a7912401e3.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-6d39136264af6513.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-2494d295a9e49f01.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-1bd94bc0a5931f31.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-af24b1ebe806a3c8.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-32bdb1831662c978.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-dc8e3506e69419c8.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-012345f38d5585e7.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui/cmp_owned" "tests/ui/cmp_owned/with_suggestion.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/cmp_owned/with_suggestion.stderr` to the actual output
--- tests/ui/cmp_owned/with_suggestion.stderr
+++ <stderr output>
+++ <stderr output>
 error: this creates an owned instance just for comparison
   --> tests/ui/cmp_owned/with_suggestion.rs:5:14
    |
 LL |         x != "foo".to_string();
-   |              ^^^^^^^^^^^^^^^^^ help: try: `"foo"`
+   |              ^^^^^^^^^^^^^^^^^ help: try: `*"foo"`
    = note: `-D clippy::cmp-owned` implied by `-D warnings`
... 4 lines skipped ...
    |
    |
 LL |         "foo".to_string() != x;
-   |         ^^^^^^^^^^^^^^^^^ help: try: `"foo"`
+   |         ^^^^^^^^^^^^^^^^^ help: try: `*"foo"`
 error: this creates an owned instance just for comparison
   --> tests/ui/cmp_owned/with_suggestion.rs:14:10
    |
    |
 LL |     x != "foo".to_owned();
-   |          ^^^^^^^^^^^^^^^^ help: try: `"foo"`
+   |          ^^^^^^^^^^^^^^^^ help: try: `*"foo"`
 error: this creates an owned instance just for comparison
   --> tests/ui/cmp_owned/with_suggestion.rs:16:10
    |
 LL |     x != String::from("foo");
 LL |     x != String::from("foo");
-   |          ^^^^^^^^^^^^^^^^^^^ help: try: `"foo"`
+   |          ^^^^^^^^^^^^^^^^^^^ help: try: `*"foo"`
 error: this creates an owned instance just for comparison
... 11 lines skipped ...
 error: aborting due to 6 previous errors
 
```

</details>

joshtriplett commented:
Some of the changes in Clippy look good; for instance:

```diff
--- a/src/tools/clippy/tests/ui/iter_overeager_cloned.fixed
+++ b/src/tools/clippy/tests/ui/iter_overeager_cloned.fixed
@@ -13,13 +13,13 @@ fn main() {
 
     let _: Option<String> = vec.iter().chain(vec.iter()).next().cloned();
 
-    let _: usize = vec.iter().filter(|x| x == &"2").count();
+    let _: usize = vec.iter().filter(|x| x == "2").count();
```

```diff
--- a/src/tools/clippy/tests/ui/op_ref.fixed
+++ b/src/tools/clippy/tests/ui/op_ref.fixed
@@ -18,7 +18,7 @@ fn main() {
     let a = "a".to_string();
     let b = "a";
 
-    if b < &a {
+    if b < a {
```

It's good that this code doesn't need to write `&"2"`.

However, some of them look bad, and seem to need a non-trivial fix in a Clippy lint:

```patch
--- a/src/tools/clippy/tests/ui/cmp_owned/with_suggestion.fixed
+++ b/src/tools/clippy/tests/ui/cmp_owned/with_suggestion.fixed
@@ -2,18 +2,18 @@
 #[allow(clippy::unnecessary_operation, clippy::no_effect, unused_must_use, clippy::eq_op)]
 fn main() {
     fn with_to_string(x: &str) {
-        x != "foo";
+        x != *"foo";
 
-        "foo" != x;
+        *"foo" != x;
```

```diff
--- a/src/tools/clippy/tests/ui/cmp_owned/with_suggestion.stderr
+++ b/src/tools/clippy/tests/ui/cmp_owned/with_suggestion.stderr
@@ -2,7 +2,7 @@ error: this creates an owned instance just for comparison
   --> tests/ui/cmp_owned/with_suggestion.rs:5:14
    |
 LL |         x != "foo".to_string();
-   |              ^^^^^^^^^^^^^^^^^ help: try: `"foo"`
+   |              ^^^^^^^^^^^^^^^^^ help: try: `*"foo"`
    |
    = note: `-D clippy::cmp-owned` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::cmp_owned)]`
```

This should not be suggesting `*"foo"`, since `"foo"` works just fine.

Could someone from @rust-lang/clippy help me figure out how to get clippy to not make that suggestion for `clippy::cmp-owned`?

joshtriplett committed:
clippy: In cmp_owned, don't use Deref via `*` if non-Deref impl exists

This avoids making a suggestion to write `*x` if `x` would also work.

joshtriplett committed:
clippy: uibless new test output

In a couple of cases, adjust the test file, because otherwise the file
produces:

```
error: failed to apply suggestions for tests/ui/iter_overeager_cloned.rs with rustfix
cannot replace slice of data that was already replaced
Add //@no-rustfix to the test file to ignore rustfix suggestions
```

joshtriplett commented:
Nevermind, I managed to figure it out.

joshtriplett commented:
@bors try

bors commented:
:hourglass: Trying commit d9721088541ee04f5cdfff2c3627314fc4113df5 with merge da4929024a7261062885cb3209c9d7fdc919d233...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d9721088541ee04f5cdfff2c3627314fc4113df5","merge_sha":"da4929024a7261062885cb3209c9d7fdc919d233"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12797480709/job/35683622592)
Build commit: da4929024a7261062885cb3209c9d7fdc919d233 (`da4929024a7261062885cb3209c9d7fdc919d233`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12797480709/job/35683622592"},"merge_sha":"da4929024a7261062885cb3209c9d7fdc919d233"} -->

joshtriplett commented:
@craterbot check

joshtriplett committed:
Add PartialEq and PartialOrd impls for comparing `String` and `&String`

This makes such comparisons more robust against the availability of
other impls.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12843850909/job/35816257819) [(plain)](https://github.com/rust-lang/rust/commit/f7320e2afe5e25e90b4fdae172b02408628125a0/checks/35816257819/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking xz2 v0.1.7
error: taken reference of right operand
  --> src/bootstrap/src/core/build_steps/clippy.rs:80:57
   |
80 |             let position = all_args.iter().position(|t| t == &rule || t == v).unwrap_or(usize::MAX);
   |                                                              |
   |                                                              help: use the right value directly: `rule`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#op_ref
```

</details>

joshtriplett commented:
I've analyzed all the crater errors:

- Some failures due to missing impls for `String` and `&String`, which I've now added.
- A failure in an old version of `sudoku`; the current version no longer has that failure.
- A failure in `koto-runtime`; I've submitted a PR at https://github.com/koto-lang/koto/pull/411
- A failure in `flexstr`; I've submitted a PR at https://github.com/nu11ptr/flexstr/pull/17
- A failure in `rbatis-codegen`; I've submitted a PR at https://github.com/rbatis/rbatis/pull/574
- A couple of other failures in a few git repos; I've submitted PRs for repos which aren't archived.

Based on that analysis, I think this is now safe to merge without causing substantial inference failures, as soon as it passes CI.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12844222333/job/35816992179) [(plain)](https://github.com/rust-lang/rust/commit/9485326c40083b7ee0a3dc40691a101738d83cb2/checks/35816992179/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.5s done
#21 DONE 39.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/inference/issue-72616.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/inference/issue-72616/issue-72616.stderr"
diff of stderr:

9    = note: multiple `impl`s satisfying `String: PartialEq<_>` found in the `alloc` crate:
10            - impl PartialEq for String;
11            - impl<'a, 'b> PartialEq<&'a str> for String;
+            - impl<'a, 'b> PartialEq<&String> for String;
12            - impl<'a, 'b> PartialEq<Cow<'a, str>> for String;
13            - impl<'a, 'b> PartialEq<str> for String;


The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args inference/issue-72616.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/inference/issue-72616.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/inference/issue-72616" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
error[E0283]: type annotations needed
##[error]  --> /checkout/tests/ui/inference/issue-72616.rs:22:37
   |
   |
LL |         if String::from("a") == "a".try_into().unwrap() {}
   |                              |
   |                              type must be known at this point
   |
   |
   = note: multiple `impl`s satisfying `String: PartialEq<_>` found in the `alloc` crate:
           - impl PartialEq for String;
           - impl<'a, 'b> PartialEq<&'a str> for String;
           - impl<'a, 'b> PartialEq<&String> for String;
           - impl<'a, 'b> PartialEq<Cow<'a, str>> for String;
           - impl<'a, 'b> PartialEq<str> for String;
   |
   |
LL |         if String::from("a") == <&str as TryInto<T>>::try_into("a").unwrap() {}
   |                                 +++++++++++++++++++++++++++++++   ~
error[E0283]: type annotations needed
##[error]  --> /checkout/tests/ui/inference/issue-72616.rs:22:37
   |
   |
LL |         if String::from("a") == "a".try_into().unwrap() {}
   |
   |
   = note: multiple `impl`s satisfying `_: TryFrom<&str>` found in the following crates: `core`, `std`:
           - impl TryFrom<&str> for std::sys_common::net::LookupHost;
           - impl<T, U> TryFrom<U> for T
             where U: Into<T>;
   = note: required for `&str` to implement `TryInto<_>`
   |
   |
LL |         if String::from("a") == <&str as TryInto<T>>::try_into("a").unwrap() {}
   |                                 +++++++++++++++++++++++++++++++   ~
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0283`.
------------------------------------------
```

</details>

joshtriplett commented:
(Currently running `./x test --bless` locally.)

joshtriplett commented:
Since new trait impls are insta-stable, this will need a libs-api FCP.

See https://github.com/rust-lang/rust/pull/135536#issuecomment-2599708371 for detailed triage of the crater run; issues in current/active projects have been addressed either with fixes to the PR or with PRs to the projects.

@rfcbot merge

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12845165354/job/35818950131) [(plain)](https://github.com/rust-lang/rust/commit/d7abdfee59990b173b300d71b1d75a377672cc59/checks/35818950131/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.private.fixed ... ok
tests/ui/ref_option/ref_option.all.fixed ... ok

FAILED TEST: tests/ui/iter_overeager_cloned.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-e60bfdd447c2b980.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-5813416a2294dc8d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-92ebd8e98d13af54.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-ddb101719ab70c6c.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-e573b0a7912401e3.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-6d39136264af6513.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-2494d295a9e49f01.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-1bd94bc0a5931f31.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-af24b1ebe806a3c8.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-32bdb1831662c978.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-d1c9d4326af97dc8.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-dc8e3506e69419c8.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-012345f38d5585e7.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/iter_overeager_cloned.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/iter_overeager_cloned.stderr` to the actual output
--- tests/ui/iter_overeager_cloned.stderr
+++ <stderr output>
+++ <stderr output>
 error: unnecessarily eager cloning of iterator items
   --> tests/ui/iter_overeager_cloned.rs:12:29
... 141 lines skipped ...
    |             help: try: `.filter(move |&(&a, b)| a == 1 && b == &target).cloned()`
+error: taken reference of right operand
+  --> tests/ui/iter_overeager_cloned.rs:50:55
+   |
+   |
+LL |         iter.cloned().filter(move |(&a, b)| a == 1 && b == &target)
+   |                                                            |
+   |                                                            help: use the right value directly: `target`
+
 error: unnecessarily eager cloning of iterator items
 error: unnecessarily eager cloning of iterator items
   --> tests/ui/iter_overeager_cloned.rs:61:13
... 4 lines skipped ...
    |                 help: try: `.filter(move |&S { a, b }| **a == 1 && b == &target).cloned()`
+error: taken reference of right operand
+  --> tests/ui/iter_overeager_cloned.rs:61:64
+   |
+   |
+LL |             iter.cloned().filter(move |S { a, b }| **a == 1 && b == &target)
+   |                                                                     |
+   |                                                                     help: use the right value directly: `target`
+
 error: unneeded cloning of iterator items
 error: unneeded cloning of iterator items
   --> tests/ui/iter_overeager_cloned.rs:65:13
... 28 lines skipped ...
    |                       help: try: `.any(|x| x.len() == 1)`
-error: aborting due to 21 previous errors
+error: aborting due to 23 previous errors
 

```

</details>

joshtriplett commented:
(clippy needs another `uibless`; working on that.)

bors commented:
:umbrella: The latest upstream changes (presumably #135947) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented:
Rebasing and reblessing.

bors commented:
:umbrella: The latest upstream changes (presumably #136209) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ibraheemdev commented on library/alloc/src/string.rs:
```diff
@@ -2530,12 +2531,51 @@ macro_rules! impl_eq {
+
+macro_rules! impl_ord {
+    ($lhs:ty, $rhs: ty) => {
+        #[stable(feature = "rust1", since = "1.0.0")]
```
Looks like some of the stability attributes were copy-pasted.

joshtriplett replied:
Fixed.

joshtriplett commented:
@ibraheemdev Does this now look good, other than waiting on FCP to complete?

ibraheemdev commented:
Yes the implementation looks good. r=me after the FCP.

riking commented on library/alloc/src/string.rs:
```diff
@@ -2530,12 +2531,51 @@ macro_rules! impl_eq {
 impl_eq! { String, str }
 impl_eq! { String, &'a str }
+impl_eq! { String, &String }
+impl_eq! { &String, str }
```
These two impls end up with incorrect stability attributes, which means Clippy msrv autodetection can't be added later.

joshtriplett replied:
That's fixable, but is that worth complicating the macro for? There are many other examples of macro-generated instances with the same discrepancy.

flip1995 commented on src/tools/clippy/tests/ui/iter_overeager_cloned.stderr:
```diff
@@ -28,6 +28,17 @@ LL |     let _: usize = vec.iter().filter(|x| x == &"2").cloned().count();
    = note: `-D clippy::redundant-clone` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::redundant_clone)]`
 
+error: taken reference of right operand
```
This lint shouldn't trigger in the test file. Please fix the two new occurrences  by removing the ref.

joshtriplett replied:
Alright, will fix.

bors commented:
:umbrella: The latest upstream changes (presumably #137752) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
